packages/backend/src/services/ProjectService.ts

```diff
@@ -58,39 +59,79 @@ export class ProjectService {
         return 'ready';
     }
 
+    async hasProject(): Promise<boolean> {
+        return this.projectModel.hasProjects();
+    }
+
     async getProject(projectUuid: string, user: SessionUser): Promise<Project> {
         // Todo: Check user has access
         const project = await this.projectModel.get(projectUuid);
         return project;
     }
 
-    async updateWarehouseConnection(
+    async create(user: SessionUser, data: CreateProject): Promise<Project> {
+        await ProjectService.testProjectAdapter(data);
+        const projectUuid = await this.projectModel.create(
+            user.organizationUuid,
+            data,
+        );
+        analytics.track({
+            event: 'project.created',
+            userId: user.userUuid,
+            properties: {
+                projectUuid,
+                projectType: data.dbtConnection.type,
+                warehouseConnectionType: data.warehouseConnection?.type,
+            },
+        });
+        await this.startAdapter(projectUuid);
+        return this.getProject(projectUuid, user);
+    }
+
+    async update(
         projectUuid: string,
         user: SessionUser,
-        data: CreateWarehouseCredentials,
-    ): Promise<Project> {
-        await this.projectModel.updateCredentials(projectUuid, data);
-        const project = await this.getProject(projectUuid, user);
+        data: UpdateProject,
+    ): Promise<void> {
+        await ProjectService.testProjectAdapter(data);
+        await this.projectModel.update(projectUuid, data);
         analytics.track({
             event: 'project.updated',
             userId: user.userUuid,
             properties: {
                 projectUuid,
-                projectType: project.dbtConnection.type,
-                warehouseConnectionType: project.warehouseConnection?.type,
+                projectType: data.dbtConnection.type,
+                warehouseConnectionType: data.warehouseConnection?.type,
             },
         });
-        return project;
+        await this.startAdapter(projectUuid);
     }
 
-    async startAdapter(projectUuid: string): Promise<ProjectAdapter> {
-        const activeAdapter = this.projectAdapters[projectUuid];
-        if (activeAdapter === undefined) {
-            const project = await this.projectModel.get(projectUuid);
-            const adapter = projectAdapterFromConfig(project.dbtConnection);
-            return adapter;
+    private static async testProjectAdapter(
+        data: UpdateProject,
+    ): Promise<void> {
+        const adapter = await projectAdapterFromConfig(data.dbtConnection);
+        if (adapter instanceof DbtLocalProjectAdapter) {
+            if (data.warehouseConnection) {
+                await adapter.updateProfile(data.warehouseConnection);
+            }
         }
-        return activeAdapter;
+        await adapter.test();
```

Comment:
Do you know what happens to the child process at this point? Does it get cleaned up correctly?

We might have to call `adapter.childProcess.kill()` depending on the adapter type